File: go/0078_subsets.go
package main

func subsets(nums []int) [][]int {
	res := [][]int{}
	backtrack([]int{}, nil, &res)
	for i, num := range nums {
		backtrack([]int{num}, nums[i+1:], &res)
	}
	return res
}

func backtrack(subset []int, nums []int, res *[][]int) {
	if len(nums) == 0 {
		*res = append(*res, subset)
		return
	}

	backtrack(subset, nil, res)
	for i, num := range nums {
		newSubset := make([]int, len(subset)+1)
		copy(newSubset[0:len(subset)], subset)
		newSubset[len(subset)] = num

		backtrack(newSubset, nums[i+1:], res)
	}
}

File: 0078_subsets.py
class Solution:
    def subsets(self, nums):
        res, nlen = [[]], len(nums)

        for i in range(nlen):
            res.append([nums[i]])
            self._subsets(nums, res, [nums[i]], i)
        return res

    def _subsets(self, nums, res, cur, prev):
        if prev + 1 == len(nums):
            return

        for i in range(prev + 1, len(nums)):
            newCur = cur.copy()
            newCur.append(nums[i])
            res.append(newCur)
            self._subsets(nums, res, newCur, i)